charon/src/ast/types/vars.rs

```diff
@@ -0,0 +1,172 @@
+///                                   Var id: 1
+/// ```
+#[derive(Debug, PartialEq, Eq, Copy, Clone, Hash, PartialOrd, Ord, Serialize, Deserialize)]
+pub enum DeBruijnVar<Bound, Free> {
+    /// A variable attached to the nth binder, counting from the inside.
+    Bound(DeBruijnId, Bound),
+    /// A variable attached to an implicit binder outside all other binders. This is not present in
+    /// translated code, and only provided as a convenience for convenient variable manipulation.
+    Free(Free),
+}
+
+// We need to manipulate a lot of indices for the types, variables, definitions, etc. In order not
+// to confuse them, we define an index type for every one of them (which is just a struct with a
+// unique usize field), together with some utilities like a fresh index generator, using the
+// `generate_index_type` macro.
+generate_index_type!(TypeVarId, "T");
+generate_index_type!(BoundRegionId, "BoundRegion");
+generate_index_type!(FreeRegionId, "FreeRegion");
+generate_index_type!(ConstGenericVarId, "Const");
+generate_index_type!(TraitClauseId, "TraitClause");
+
+/// A type variable in a signature or binder.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Drive, DriveMut)]
+pub struct TypeVar {
+    /// Index identifying the variable among other variables bound at the same level.
+    pub index: TypeVarId,
+    /// Variable name
+    pub name: String,
+}
+
+/// A region variable in a signature or binder.
+#[derive(
+    Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Hash, PartialOrd, Ord, Drive, DriveMut,
+)]
+pub struct RegionVar {
+    /// Index identifying the variable among other variables bound at the same level.
+    pub index: BoundRegionId,
+    /// Region name
+    pub name: Option<String>,
+}
+
+/// A const generic variable in a signature or binder.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Drive, DriveMut)]
+pub struct ConstGenericVar {
+    /// Index identifying the variable among other variables bound at the same level.
+    pub index: ConstGenericVarId,
+    /// Const generic name
+    pub name: String,
+    /// Type of the const generic
+    pub ty: LiteralTy,
+}
+
+/// A trait predicate in a signature, of the form `Type: Trait<Args>`. This functions like a
+/// variable binder, to which variables of the form `TraitRefKind::Clause` can refer to.
+#[derive(Debug, Clone, Serialize, Deserialize, Drive, DriveMut)]
+pub struct TraitClause {
+    /// Index identifying the clause among other clauses bound at the same level.
+    pub clause_id: TraitClauseId,
+    // TODO: does not need to be an option.
+    pub span: Option<Span>,
+    /// Where the predicate was written, relative to the item that requires it.
+    #[charon::opaque]
+    pub origin: PredicateOrigin,
+    /// The trait that is implemented.
+    #[charon::rename("trait")]
+    pub trait_: PolyTraitDeclRef,
+}
+
+pub type RegionDbVar = DeBruijnVar<BoundRegionId, FreeRegionId>;
```

Comment:
I just realized I'm not super happy with the name `DeBruijnVar`: to me a de Bruijn var is necessarily a bound var, while here it actually refers to a bound var or a free var. I'm aware the name `Var` is already taken. Maybe we should use `VarDecl` (for the current `Var` type) and `Var` (for the current `DeBruijnVar`)?